odoo/modules/packages.py

```diff
@@ -0,0 +1,148 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+""" Modules dependency graph. """
+
+import logging
+# Deprecated since version 3.9:
+# https://docs.python.org/3.9/library/typing.html#typing.Iterable
+from typing import List, Dict, Iterable, Iterator
+
+import odoo
+import odoo.tools as tools
+
+_logger = logging.getLogger(__name__)
+
+
+class Package:
+    """
+    Python package for the Odoo module with the same name
+    """
+    def __init__(self, name: str) -> None:
+        # manifest data
+        self.name = name
+        self.manifest = manifest = odoo.modules.module.get_manifest(name) or {}
+
+        # ir_module_module data             # column_name
+        self.id = None                      # id
+        self.state = 'uninstalled'          # state
+        self.dbdemo = False                 # demo
+        self.installed_version = None       # latest_version
+
+        # info for upgrade
+        self.upgrade_end = False            # if the package is need to upgrade end-xxx
+        self.old_version = None             # the version when added to packages
+
+        # dependencies
+        self.depends = set(manifest.get('depends', ()))
+        self.dependencies: List[Package] = []
+
+        self._depth_loop = False            # detects dependency loops
+
+    @tools.lazy_property
+    def depth(self) -> int:
+        """ Return the longest distance from self to module 'base' along dependencies. """
+        if self._depth_loop:
+            raise Exception('depends_loop')
+        self._depth_loop = True
+        if self.depends:
+            result = max(package.depth for package in self.dependencies) + 1
+        else:
+            assert self.name == 'base'
+            result = 0
+        self._depth_loop = False
+        return result
+
+    def sorting_key(self):
+        return (self.depth, self.name)
```

Comment:
It can actually be a lambda, and shoven directly into the call to `sorted()`:
```py
return iter(sorted(self._packages.values(), key=lambda package: (package.depth, package.name)))
```